src/xds/ads_client.rs

```diff
@@ -131,145 +90,192 @@ impl AdsClient {
-                    // We have previously validated that a config provides at least one
-                    // server address so this default value shouldn't be necessary.
-                    .unwrap_or_else(|| "127.0.0.1:18000".into());
-                next_server_index += 1;
-                server_addr
+            match error {
+                RpcSessionError::NonRecoverable(msg, err) => {
+                    error!(log, "{}\n{}", msg, err);
+                    RetryPolicy::Break
+                }
+
+                RpcSessionError::InitialConnect(ref err) => {
+                    error!(log, "Unable to connect to the XDS server"; "error" => %err);
+
+                    // Do not retry if this is an invalid URL error that we cannot recover from.
+                    // Need to use {:?} as the Display output only returns 'transport error'
+                    let err_description = format!("{:?}", err);
+                    if err_description.to_lowercase().contains("invalid url") {
+                        RetryPolicy::Break
+                    } else {
+                        RetryPolicy::Delay(backoff.delay(attempt, &error))
+                    }
+                }
+
+                RpcSessionError::Receive(ref status) => {
+                    error!(log, "Failed to receive response from XDS server"; "status" => #?status);
+                    RetryPolicy::Delay(backoff.delay(attempt, &error))
+                }
+            }
+        });
+
+        let session_shutdown_rx = shutdown_rx.clone();
+        let handle = tryhard::retry_fn(|| {
+            let (discovery_req_tx, discovery_req_rx) =
+                mpsc::channel::<DiscoveryRequest>(UPDATES_CHANNEL_BUFFER_SIZE);
+            let cluster_manager = ClusterManager::new(
+                log.clone(),
+                cluster_updates_tx.clone(),
+                discovery_req_tx.clone(),
+            );
+            let listener_manager =
+                ListenerManager::new(log.clone(), listener_manager_args.clone(), discovery_req_tx);
+
+            let resource_handlers = ResourceHandlers {
+                cluster_manager,
+                listener_manager,
             };
 
-            let args = RpcSessionArgs {
+            RpcSession {
+                discovery_req_rx,
                 log: log.clone(),
                 metrics: metrics.clone(),
-                server_addr: server_addr.clone(),
                 node_id: node_id.clone(),
+                addr: server_iter
+                    .next()
+                    .map(|server| server.address.to_string())
+                    .unwrap(),
                 resource_handlers,
-                backoff,
-                discovery_req_rx: &mut discovery_req_rx,
-                shutdown_rx: shutdown_rx.clone(),
+                shutdown_rx: session_shutdown_rx.clone(),
+            }
+            .run()
+        })
+        .with_config(retry_config);
+
+        tokio::select! {
```

Comment:
I'll defer the majority of the review to @iffyio who knows this code section far better than I - but I will definitely say that this section is far more clear 👍🏻